src/rhsmlib/file_monitor.py

```diff
@@ -0,0 +1,220 @@
+            timestamp = self.update_timestamp(dw)
+            if dw.timestamp != timestamp:
+                changed_dws.add(dw)
+                dw.timestamp = timestamp
+        return changed_dws
+
+    def loop(self, callback=None):
+        """
+        initializes timestamps for each dir watch in dir watch list, sets callback,
+        notifies dir watch if it has changed
+        :param callback: callback function to be called at end of each iteration of the loop
+        :return:
+        """
+        for dw in self.dir_watches:
+            dw.timestamp = self.update_timestamp(dw)
+        if callback is None:
+            callback = end_loop_cb
+        while not self.shouldStop:
+            self.update()
+            time.sleep(2.0)
+            if callback:
+                self.shouldStop = callback(self)
+        self.stop()
+
+
+class InotifyFilesystemWatcher(FilesystemWatcher):
+    def __init__(self, dir_watches):
+        """
+        Filesystem watcher if pyinotify is configured and available
+        loop function will override parent class loop function
+        :param dir_watches: list of directories to watch
+        """
+        FilesystemWatcher.__init__(self, dir_watches)
+        self.watchManager = None
+        self.notifier = None
+
+    def loop(self, callback=None):
+        """
+        sets up watch manager, notifier, adds watches to watch manager, and starts loop
+        :param callback: callback method to be called at the end of each iteration of the loop
+        :return:
+        """
+        self.watchManager = pyinotify.WatchManager()
+        self.notifier = pyinotify.Notifier(self.watchManager, self.handleEvent)
+        self.add_watches()
+        self.notifier.loop(callback)
+
+    def handleEvent(self, event):
+        """
+        default process function for pyinotify notifier
+        :param event: pyinotify Event object, has path and mask of flags representing file modification
+        :return:
+        """
+        for dir_watch in self.dir_watches:
+            if dir_watch.pathsMatch(event.path) and dir_watch.fileModified(event.mask):
+                dir_watch.notify()
+
+    def add_watches(self):
+        """
+        adds watches to the watch manager
+        :return:
+        """
+        for dir_watch in self.dir_watches:
+            self.watchManager.add_watch(dir_watch.path, dir_watch.mask, do_glob=dir_watch.glob)
+
+
+class DirectoryWatch(object):
+    def __init__(self, path, callbacks, glob=False):
+        self.IN_CREATE = 0x00000100
+        self.IN_DELETE = 0x00000200
+        self.IN_MODIFY = 0x00000002
+
+        self.path = path
+        self.timestamp = None
+        self.glob = glob
+        self.callbacks = callbacks
+        self.mask = self.IN_CREATE | self.IN_DELETE | self.IN_MODIFY
+
+    def notify(self):
+        """
+        calls all callbacks associated with dir watch
+        :return:
+        """
+        for cb in self.callbacks:
+            cb()
```

Comment:
You can try to call anything. E.g. None. When callback function raise any exception, then server will stop to work.